src/python/pants/backend/go/util_rules/coverage_html.py

```diff
@@ -0,0 +1,294 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import io
+import math
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import Sequence
+
+import chevron
+
+from pants.backend.go.util_rules.coverage import GoCoverMode
+from pants.backend.go.util_rules.coverage_profile import (
+    GoCoverageBoundary,
+    GoCoverageProfile,
+    parse_go_coverage_profiles,
+)
+from pants.engine.fs import DigestContents
+from pants.engine.internals.native_engine import Digest
+from pants.engine.internals.selectors import Get
+from pants.engine.rules import collect_rules, rule
+
+# Adapted from Go toolchain.
+# See https://github.com/golang/go/blob/a0441c7ae3dea57a0553c9ea77e184c34b7da40f/src/cmd/cover/html.go
```

Comment:
Ugh, this is pretty unfortunate to have to do. Is there no way to invoke the go tool to convert an existing raw coverage file to html (as we do for python coverage), rather than emulate what that does?

Reply:
There isn't because `go tool cover` attempts to find the source files on its own using `go list`. See https://github.com/golang/go/blob/a0441c7ae3dea57a0553c9ea77e184c34b7da40f/src/cmd/cover/func.go#L200-L222.

We have had to engineer the Go backend to avoid `go list` due to it needing, among other things, all transitive third-party dependencies available to it when analyzing first-party sources. Thus, the use of `go list` by `go tool cover` in this case means we cannot use `go tool cover` to generate the HTML.

Reply:
Added that explanation as a comment.

Reply:
This will introduce a huge maintenance challenge. How do we keep this in sync with changes to `go tool cover`? 

Wondering if there is a way to "monkeypatch" via some shim code that delegates to html.go ?

Reply:
A "monkeypatch" would need to somehow intercept the `go list` invocation invoked by `go tool cover`. Note in the [source code though](https://github.com/golang/go/blob/a0441c7ae3dea57a0553c9ea77e184c34b7da40f/src/cmd/cover/func.go#L202), `go tool cover` specifically uses its own `GOROOT` to find and invoke `go list`.  Trying to intercept `go tool cover` in this manner is extremely brittle if it can be accomplished as it is.

As for delegating to html.go, that would involve cloning the upstream `go` repo somehow. The relevant functions are not exported from its package and so they cannot be consumed as a third-party library.

In any event, this begs the question: Should Pants provide the _same_ experience or an _equivalent_ experience? Here, we should provide _equivalent_ even if it could diverge in the future.

The HTML report generated by this PR matches existing `go tool cover` behavior without any brittle trickery to obtain the _same_ experience. If that experience diverges in the future, the report generated by this PR is still useful and valid, regardless of what the upstream does.

Reply:
Side note: Bazel `rules_go` doesn't produce HTML reports; it actually generate portable `lcov` reports which the user can then visualize with lcov-related tooling. The goal of this PR was to provide a semi-useful report that is better than trying to read the raw coverage report.

But Pants standardizing on lcov for an option in all language backends is a separate matter.

Reply:
The existing `go tool cover` behavior is still subject to bugfixes, which we would not pick up.

Also, assuming `go tool cover` behavior changes in the future, wouldn't it be weird for users to see older behavior when they are otherwise using a newer go? 

I like the idea of pants standardizing on lcov, and we can than have a plugin for lcov-related tooling that works across languages.